src/gnatss/ops/data.py

```diff
@@ -231,6 +231,37 @@ def get_data_inputs(all_observations: pd.DataFrame) -> NumbaList:
     return data_inputs
 
 
+def prefilter_replies(
+    all_observations: pd.DataFrame,
+    num_transponders: int,
+) -> pd.DataFrame:
+    """
+    Remove pings that do receive replies from each
+    transponder in the array.
+
+    Parameters
+    ----------
+    all_observations : pd.DataFrame
+        The original observations that include every ping and reply
+    num_transponders : int
+        The number of transponders in the array
+
+    Returns
+    -------
+    pd.DataFrame
+        The observations where the number of replies equal the
+        number of transponders
+    """
+    # Get value counts for transmit times
+    time_counts = all_observations[constants.DATA_SPEC.tx_time].value_counts()
+
+    return all_observations[
+        all_observations[constants.DATA_SPEC.tx_time].isin(
+            time_counts[time_counts == num_transponders].index
```

Comment:
Is this filtering still works with the current structure of the observations dataframe? This looks like from a previous implementation with the structure being:

```
transmit_time, reply_t1, reply_t2, reply_t3
1, 1, 2, 3
```

rather than a current one of 

```
reply_times,transmit_times,transponder
1, 1, t1
2, 1, t2
3, 1, t3
```

Please correct me if I'm wrong

Reply:
So I actually had some trouble figuring out what was in the dataframe just from reading through the source code alone so I copied the formulation currently being used in `get_data_inputs()`.

I have tested this fix with a data set that replicates the bug in #296 and it works. However, I can't actually answer your question about the dataframe structure, which may be dovetailing with another issue I've found. (Short version: the i/o will fail if not given a path to tt files in the config file, even if you are just running the solver on previously generated gps_solutions.csv data. It also reads in all posfilter input even if --from-cache is set.)

Reply:
More context. Here is the output when processing a bad data set with the current 0.2.0 version:

```
gnatss run --solver --distance-limit 150 NDP1_2024_config.yaml 
Starting GNATSS ...
Gathering travel_times at /Users/johndesanto/Documents/gnatss_testing/**/WG*/pxp_tt
Gathering sound_speed at /Users/johndesanto/Documents/gnatss_testing/ctd/NDP1_Ch_Mi_2022_fit
Gathering gps_solution at /Users/johndesanto/Documents/gnatss_testing/output/gps_solution.csv
Loading travel_times from /Users/johndesanto/Documents/gnatss_testing/**/WG*/pxp_tt
Loading sound_speed from /Users/johndesanto/Documents/gnatss_testing/ctd/NDP1_Ch_Mi_2022_fit
Loading gps_solution from /Users/johndesanto/Documents/gnatss_testing/output/gps_solution.csv
Loading deletions from /Users/johndesanto/Documents/gnatss_testing/output/
Preprocessing Travel Times Data
Finished Preprocessing Travel Times Data
Computing harmonic mean...
lat=44.81792965 lon=-125.12664945 height=-1829.645 alt=-1829.645 internal_delay=0.08 sv_mean=1482.944 pxp_id='NDP1-2' azimuth=-107.66 elevation=41.67
lat=44.8423252 lon=-125.13482028 height=-1830.76 alt=-1830.76 internal_delay=0.36 sv_mean=1482.944 pxp_id='NDP1-3' azimuth=134.54 elevation=41.59
lat=44.83268136 lon=-125.0997949 height=-1832.822 alt=-1832.822 internal_delay=0.64 sv_mean=1482.944 pxp_id='NDP1-1' azimuth=3.06 elevation=41.67
Finished computing harmonic mean
Filtering out data outside of distance limit...
Preparing data inputs...
Perform solve...
--- 2037 epochs, 6104 measurements ---
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/johndesanto/miniforge3/envs/gnatss/lib/python3.10/site-packages/gnatss/cli.py:100 in run  │
│                                                                                                  │
│    97 │   │   skip_posfilter = not posfilter                                                     │
│    98 │   │   skip_solver = not solver                                                           │
│    99 │                                                                                          │
│ ❱ 100 │   run_gnatss(                                                                            │
│   101 │   │   config_yaml=config_yaml,                                                           │
│   102 │   │   distance_limit=distance_limit,                                                     │
│   103 │   │   residual_limit=residual_limit,                                                     │
│                                                                                                  │
│ /Users/johndesanto/miniforge3/envs/gnatss/lib/python3.10/site-packages/gnatss/main.py:110 in     │
│ run_gnatss                                                                                       │
│                                                                                                  │
│   107 │   │   data_dict = run_posfilter(config, data_dict)                                       │
│   108 │                                                                                          │
│   109 │   if config.solver and not skip_solver:                                                  │
│ ❱ 110 │   │   data_dict = run_solver(config, data_dict, return_raw=return_raw)                   │
│   111 │   │   # Write out to residuals.csv file                                                  │
│   112 │   │   to_file(config, data_dict, "residuals", CSVOutput.residuals)                       │
│   113                                                                                            │
│                                                                                                  │
│ /Users/johndesanto/miniforge3/envs/gnatss/lib/python3.10/site-packages/gnatss/solver/run.py:33   │
│ in run_solver                                                                                    │
│                                                                                                  │
│   30 │   all_epochs = get_all_epochs(all_observations)                                           │
│   31 │                                                                                           │
│   32 │   twtt_model = config.solver.twtt_model                                                   │
│ ❱ 33 │   process_data, is_converged = prepare_and_solve(all_observations, config, twtt_model=    │
│   34 │                                                                                           │
│   35 │   if is_converged:                                                                        │
│   36 │   │   _print_final_stats(config.solver.transponders, process_data)                        │
│                                                                                                  │
│ /Users/johndesanto/miniforge3/envs/gnatss/lib/python3.10/site-packages/gnatss/solver/utilities.p │
│ y:414 in prepare_and_solve                                                                       │
│                                                                                                  │
│   411 │   │   process_dict[n_iter] = {"transponders_xyz": transponders_xyz}                      │
│   412 │   │                                                                                      │
│   413 │   │   # Perform solving                                                                  │
│ ❱ 414 │   │   all_results = perform_solve(                                                       │
│   415 │   │   │   data_inputs,                                                                   │
│   416 │   │   │   transponders_mean_sv,                                                          │
│   417 │   │   │   transponders_xyz,                                                              │
│                                                                                                  │
│ /Users/johndesanto/miniforge3/envs/gnatss/lib/python3.10/site-packages/gnatss/solver/solve.py:43 │
│ in _calc_tr_vectors                                                                              │
│                                                                                                  │
│    40 │   # Ensure transponders xyz and reply xyz have the same shape                            │
│    41 │   m, n = transponders_xyz.shape                                                          │
│    42 │   o, p = reply_xyz.shape                                                                 │
│ ❱  43 │   assert (m, n) == (o, p), f"Mismatch shape found: ({m},{n}) != ({o},{p})"               │
│    44 │                                                                                          │
│    45 │   transmit_vectors = transponders_xyz - transmit_xyz                                     │
│    46 │   reply_vectors = transponders_xyz - reply_xyz                                           │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AssertionError: Mismatch shape found: (3,3) != (1,3)
```
Note this replicates Issue #296.

Reply:
Here is the output with the bug fix:

```
gnatss run --solver --distance-limit 150 NDP1_2024_config.yaml 
Starting GNATSS ...
Gathering travel_times at /Users/johndesanto/Documents/gnatss_testing/**/WG*/pxp_tt
Gathering gps_solution at /Users/johndesanto/Documents/gnatss_testing/output/gps_solution.csv
Gathering sound_speed at /Users/johndesanto/Documents/gnatss_testing/ctd/NDP1_Ch_Mi_2022_fit
Loading travel_times from /Users/johndesanto/Documents/gnatss_testing/**/WG*/pxp_tt
Loading gps_solution from /Users/johndesanto/Documents/gnatss_testing/output/gps_solution.csv
Loading sound_speed from /Users/johndesanto/Documents/gnatss_testing/ctd/NDP1_Ch_Mi_2022_fit
Loading deletions from /Users/johndesanto/Documents/gnatss_testing/output/
Preprocessing Travel Times Data
Finished Preprocessing Travel Times Data
Computing harmonic mean...
lat=44.81792965 lon=-125.12664945 height=-1829.645 alt=-1829.645 internal_delay=0.08 sv_mean=1482.944 pxp_id='NDP1-2' azimuth=-107.66 elevation=41.67
lat=44.8423252 lon=-125.13482028 height=-1830.76 alt=-1830.76 internal_delay=0.36 sv_mean=1482.944 pxp_id='NDP1-3' azimuth=134.54 elevation=41.59
lat=44.83268136 lon=-125.0997949 height=-1832.822 alt=-1832.822 internal_delay=0.64 sv_mean=1482.944 pxp_id='NDP1-1' azimuth=3.06 elevation=41.67
Finished computing harmonic mean
Filtering out data outside of distance limit...
Preparing data inputs...
Perform solve...
--- 2031 epochs, 6093 measurements ---
After iteration: 1, rms residual = 98.07 cm, error factor = 6.902
NDP1-2
D_x = 3.907973e-01 m, Sigma(x) = 1.581696e-03 m
D_y = 2.525879e-01 m, Sigma(y) = 1.579056e-03 m
D_z = -4.369847e-01 m, Sigma(z) = 1.577894e-03 m
NDP1-3
D_x = 3.907973e-01 m, Sigma(x) = 1.581696e-03 m
D_y = 2.525879e-01 m, Sigma(y) = 1.579056e-03 m
D_z = -4.369847e-01 m, Sigma(z) = 1.577894e-03 m
NDP1-1
D_x = 3.907973e-01 m, Sigma(x) = 1.581696e-03 m
D_y = 2.525879e-01 m, Sigma(y) = 1.579056e-03 m
D_z = -4.369847e-01 m, Sigma(z) = 1.577894e-03 m

After iteration: 2, rms residual = 30.34 cm, error factor = 2.135
NDP1-2
D_x = 7.295596e-06 m, Sigma(x) = 1.581700e-03 m
D_y = -4.974752e-05 m, Sigma(y) = 1.579059e-03 m
D_z = 3.672391e-05 m, Sigma(z) = 1.577896e-03 m
NDP1-3
D_x = 7.295596e-06 m, Sigma(x) = 1.581700e-03 m
D_y = -4.974752e-05 m, Sigma(y) = 1.579059e-03 m
D_z = 3.672391e-05 m, Sigma(z) = 1.577896e-03 m
NDP1-1
D_x = 7.295596e-06 m, Sigma(x) = 1.581700e-03 m
D_y = -4.974752e-05 m, Sigma(y) = 1.579059e-03 m
D_z = 3.672391e-05 m, Sigma(z) = 1.577896e-03 m

---- FINAL SOLUTION ----
NDP1-2
x = -2606818.7046 +/- 1.581700e-03 m del_e = 0.1743 +/- 1.581447e-03 m
y = -3705463.3795 +/- 1.579059e-03 m del_n = -0.0059 +/- 1.582987e-03 m
z = 4471710.2056 +/- 1.577896e-03 m del_u = -0.6141 +/- 1.57421e-03 m
Lat. = 44.817929597184566 deg, Long. = -125.12664724583038, Hgt.msl = -1830.2590579230641 m
NDP1-3
x = -2606247.0128 +/- 1.581700e-03 m del_e = 0.1742 +/- 1.581446e-03 m
y = -3703528.3343 +/- 1.579059e-03 m del_n = -0.0056 +/- 1.582988e-03 m
z = 4473631.5288 +/- 1.577896e-03 m del_u = -0.6141 +/- 1.574209e-03 m
Lat. = 44.84232514969592 deg, Long. = -125.13481807567847, Hgt.msl = -1831.3740779860077 m
NDP1-1
x = -2604416.0665 +/- 1.581700e-03 m del_e = 0.1745 +/- 1.581447e-03 m
y = -3705737.7294 +/- 1.579059e-03 m del_n = -0.0058 +/- 1.582987e-03 m
z = 4472870.3395 +/- 1.577896e-03 m del_u = -0.614 +/- 1.574209e-03 m
Lat. = 44.832681308089015 deg, Long. = -125.09979269270931, Hgt.msl = -1833.4360014477363 m
------------------------

There are 0 outliers found during this run.

Exporting residuals to residuals.csv ...
Successfully exported data to /Users/johndesanto/Documents/gnatss_testing/output/residuals.csv
Exporting outliers to outliers.csv ...
No data found for outliers, skipping export!
Exporting distance_from_center to dist_center.csv ...
Successfully exported data to /Users/johndesanto/Documents/gnatss_testing/output/dist_center.csv
Exporting process_dataset to process_dataset.nc ...
Successfully exported data to /Users/johndesanto/Documents/gnatss_testing/output/process_dataset.nc
Successfully exported data to /Users/johndesanto/Documents/gnatss_testing/output/residuals.png
Successfully exported data to /Users/johndesanto/Documents/gnatss_testing/output/residuals_enu_components.png
Finished GNATSS.
```
Note that the fixed version operates on 2031 epochs instead of 2037 epochs because of the prefilter.

Reply:
>(Short version: the i/o will fail if not given a path to tt files in the config file, even if you are just running the solver on previously generated gps_solutions.csv data. It also reads in all posfilter input even if --from-cache is set.)

Gotcha, there might be some checks that are missing then to ensure that tt files doesn't get read in again. Okay, yea I don't remember the data structure format anymore at this point 😬 But if it works, then awesome! Though I suggest exploring a bit to get an understanding of the data structure.